sql: update stats for descriptor table

rafiss opened the pull request:
This is needed in order to enforce the
sql.schema.approx_max_object_count cluster setting, which relies on optimizer table statistics to find the count.

Since the schemachanger uses the KV API to write to the descriptor table, we need to explicitly notify the stats refresher when it should compute new stats for the table.

informs: https://github.com/cockroachdb/cockroach/issues/148286
Release note: None

###  schematelemetry: update object count gauge using table stats

In order to support larger scales of object counts, we switch away from
using a full table scan on system.descriptors in order to update the
object count gauge.

Instead, we use table stats on system.descriptor now. The
schematelemetry job is updated so it notifies the stats refresher to
keep the stats up to date. Note that the stats refresher is also
notified with a partial count already.

Epic CRDB-48806
Release note: None

rafiss committed:
sql: update stats for descriptor table

This is needed in order to enforce the
sql.schema.approx_max_object_count cluster setting, which relies on
optimizer table statistics to find the count.

Since the schemachanger uses the KV API to write to the descriptor
table, we need to explicitly notify the stats refresher when it should
compute new stats for the table.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154711)
<!-- Reviewable:end -->

rafiss committed:
schematelemetry: update object count gauge using table stats

In order to support larger scales of object counts, we switch away from
using a full table scan on system.descriptors in order to update the
object count gauge.

Instead, we use table stats on system.descriptor now. The
schematelemetry job is updated so it notifies the stats refresher to
keep the stats up to date. Note that the stats refresher is also
notified with a partial count already.

Release note: None

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

One suggestion for the test

> @fqazi reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154711#-:-OaeS_GA6bpU9vD49Qwu:b-lg09qv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/schemachanger_test.go` line 1245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154711#-OaeTRYh8pTqeRnsO5ns:-OaeTRYh8pTqeRnsO5nt:bps5hs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2905b2137118ccbcef2c4a953712f4a735a536d8/pkg/sql/schemachanger/schemachanger_test.go#L1245)):*
> ```Go
> 						// Haven't hit the limit yet, keep trying.
> 						return errors.Errorf("created %d %ss without hitting limit (max=%d)", objNum, objectType, maxObjects)
> 					}, 5*time.Minute)
> ```

We could set `sql.stats.automatic_collection.min_stale_row` to a lower value, which might speed this test up?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154711)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/schemachanger_test.go` line 1245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154711#-OaeTRYh8pTqeRnsO5ns:-OaeYjp51JTSstP0b7NQ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2905b2137118ccbcef2c4a953712f4a735a536d8/pkg/sql/schemachanger/schemachanger_test.go#L1245)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We could set `sql.stats.automatic_collection.min_stale_row` to a lower value, which might speed this test up?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>:

> @fqazi reviewed 2 of 2 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154711#-:-OaeaBSaDbluyiMVonYz:b-uwkovn)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schema_changer_metrics.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/154711#-Oaea8H32wH7CakqnbeB:-Oaea8H32wH7CakqnbeC:b2fmk2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a11aef2e29cc3aa74b16e8484bc7946cc8b4fc18/pkg/sql/schema_changer_metrics.go#L52)):*
> ```Go
> // UpdateDescriptorCount updates our sql.schema_changer.object_count gauge with
> // a fresh count of objects in the system.descriptor table.
> func UpdateDescriptorCount(
> ```

This is a lot cleaner, nice!


<!-- Sent from Reviewable.io -->
